### PR TITLE
Add check for current charter.

### LINKED ISSUE #782: If the document is on REC-track, make sure the WG is chartered to produce REC-track documents

### LINKED ISSUE #1319: Joint-publication with different version of Patent Policy.
Currently, Pubrules only take the Patent Policy link from **the first WG/IG** (as deliverer) mentioned in the document.
https://github.com/w3c/specberus/blob/3d21736b319d3077b99f60099820d99f2e5c3604/lib/rules/metadata/patent-policy.js#L17-L20

Although it can be really rare, but it could be possible for **groups publishing a spec jointly choose different versions of Patent Policy.** 

@plehegar @wseltzer any opinions?
- Shall we allow this case? (In echidna or manual publication)
- Do groups' PP versions have to be aligned to publish a spec?
- Will it cause any impact on legal stuff?

### PR DESCRIPTION
**Pull request by jennyliang220**
resolves #782 , resolves #1319

### CODE REVIEW

File: lib/validator.js [hunk 1]
```diff
@@ -506,6 +506,66 @@ Specberus.prototype.getDelivererIDsNote = function () {
...
+                ) {
+                    patentPolicy.push(groupCharter['patent-policy']);
+                    currentCharters.push(groupCharter.uri);
+                }
+            });
+        });
+    }
+
+    this.charters = currentCharters;
+    [this.patentPolicy] = patentPolicy;
```
Comment (deniak) [inline lib/validator.js:562]: If there's a joint publication, we should check that all the deliverers are operating under the same patent policy.
With the current code, there will be no error if that's not the case and only the first PP will be returned.

**Review by deniak — CHANGES_REQUESTED** [review]
`lib/rules/sotd/charter-disclosure.js` should probably be renamed to `lib/rules/sotd/charter-disclosure.js` as the "disclosure" part is not checked for documents other than IG Notes

File: lib/validator.js [hunk 2]
```diff
@@ -506,6 +506,66 @@ Specberus.prototype.getDelivererIDsNote = function () {
...
+
+    if (deliverers.length) {
+        w3capi.apiKey = process.env.W3C_API_KEY;
+
+        const delivererPromises = [];
+
+        const { TAG_ID } = util;
+        // Get charters data from W3C API
+        // deliverers.forEach is for joint publication.
+        deliverers.forEach((deliverer) => {
```
[the remark this answers is not in the document]
Comment (jennyliang220) [reply lib/validator.js:531]: This was not configured in prettier before, the [default config is to have the parentheses](https://prettier.io/docs/en/options.html#arrow-function-parentheses). So the parentheses are necessary. I don't have a preference, will add  the config to the project.

**Comment by jennyliang220** [comment]
> `lib/rules/sotd/charter-disclosure.js` should probably be renamed to `lib/rules/sotd/charter-disclosure.js` as the "disclosure" part is not checked for documents other than IG Notes

Hi @deniak , is `lib/rules/sotd/charter.js` what you meant?

**Comment by deniak** [comment]
> > `lib/rules/sotd/charter-disclosure.js` should probably be renamed to `lib/rules/sotd/charter-disclosure.js` as the "disclosure" part is not checked for documents other than IG Notes
> 
> Hi @deniak , is `lib/rules/sotd/charter.js` what you meant?

Yes, that's more generic.

File: lib/rules/sotd/pp.js [hunk 3]
```diff
@@ -151,6 +151,17 @@ exports.check = async function (sr, done) {
     } else if (sr.config.patentPolicy === 'pp2004') {
         patent = pp2017;
     }
+
+    // Make sure the pp version is consitant between the UI or the one defined in the charter.
+    if (!patentPolicies[0]) console.log('\n\n no pp!');
```
Comment (deniak) [inline lib/rules/sotd/pp.js:156]:
```suggestion
```

File: lib/validator.js [hunk 4]
```diff
@@ -552,21 +544,37 @@ Specberus.prototype.getChartersAndPolicy = async function () {
...
+    return charters;
+};
+
+Specberus.prototype.getPatentPolicies = async function () {
+    if (undefined !== this.patentPolicies) {
+        return this.patentPolicies;
+    }
+    const chartersData = await this.getChartersData();
+    const patentPolicies = [];
+    chartersData.every(charterData =>
```
Comment (deniak) [inline lib/validator.js:573]: you are mistaking `every` with `forEach`. [`every`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/every) should be use to test all elements of the array.
```suggestion
    chartersData.forEach(charterData =>
```

**Review by deniak — CHANGES_REQUESTED** [review]

File: lib/l10n-en_GB.js [hunk 5]
```diff
@@ -215,18 +215,21 @@ exports.messages = {
     'sotd.pp.no-errata':
         'No errata link was found in the patent policy section.',
     'sotd.pp.no-response': 'The connection to the API failed: ${error}',
-    // sotd/charter-disclosure
-    'sotd.charter-disclosure.no-group': 'Unable to find the deliverer id.',
-    'sotd.charter-disclosure.text-not-found':
+    'sotd.pp.joint-different-pp':
+        'This document is published by multiple groups using different versions of Patent Policy. See requirement details at <a href="https://github.com/w3c/specberus/issues/1319#issuecomment-878619033">this GitHub Issue</a>.',
```
Comment (deniak) [inline lib/l10n-en_GB.js:219]: I don't think we should link to a comment which can be edited in the future.

```suggestion
        'This document is published by multiple groups operating under different versions of the Patent Policy. A document should be published under one and only one Patent Policy, to which all participants in contributing groups have agreed.',
```

File: lib/l10n-en_GB.js [hunk 6]
```diff
@@ -215,18 +215,21 @@ exports.messages = {
...
+    'sotd.pp.wrong-pp-from-charter':
+        'The Working Group delivering the document operates under the patent policy "${pp_charter}", but the check is against "${pp_config}". Please change the UI or query parameter to match the charter\'s version.',
```
Comment (deniak) [inline lib/l10n-en_GB.js:221]:
```suggestion
        'The Working Group delivering the document operates under the patent policy "${pp_charter}", but the check is against "${pp_config}". Please select the right patent policy to match the charter\'s version.',
```

File: lib/l10n-en_GB.js [hunk 7]
```diff
@@ -215,18 +215,21 @@ exports.messages = {
...
+    // sotd/charter
+    'sotd.charter.no-group':
+        'Unable to find the group id (in data-deliverer for Note-Track document, and in Status of This Document for REC-Track document).',
```
Comment (deniak) [inline lib/l10n-en_GB.js:224]:
```suggestion
        'Unable to find the group(s) ID(s) (in the <code>data-deliverer</code> attribute for Note-Track documents, and in the "Status of This Document" for REC-Track documents).',
```

File: lib/validator.js [hunk 8]
```diff
@@ -489,23 +493,93 @@ Specberus.prototype.getDelivererIDs = async function () {
...
+    return chartersData;
+};
+
+Specberus.prototype.getCharters = async function () {
+    if (undefined !== this.charters) {
+        return this.charters;
+    }
+    const chartersData = await this.getChartersData();
+    const charters = [];
+    chartersData.every(charterData => charters.push(charterData.uri));
```
Comment (deniak) [inline lib/validator.js:565]:
```suggestion
    chartersData.forEach(charterData => charters.push(charterData.uri));
```

File: lib/validator.js [hunk 9]
```diff
@@ -489,23 +493,93 @@ Specberus.prototype.getDelivererIDs = async function () {
...
+    const chartersData = [];
+
+    if (deliverers.length) {
+        w3capi.apiKey = process.env.W3C_API_KEY;
+        const delivererPromises = [];
+        const { TAG_ID } = util;
+        // Get charters data from W3C API
+        // deliverers.forEach is for joint publication.
+        deliverers.forEach(deliverer => {
+            // Skip finding charter for TAG group.
```
Comment (deniak) [inline lib/validator.js:527]:
```suggestion
            // Skip finding charter for the TAG which doesn't have any charter
```

File: lib/validator.js [hunk 10]
```diff
@@ -489,23 +493,93 @@ Specberus.prototype.getDelivererIDs = async function () {
...
+                        .group(deliverer)
+                        .charters()
+                        .fetch({ embed: true }, (err, charters) => {
+                            resolve(charters);
+                        });
+                })
+            );
+        });
+        groupsCharters = await Promise.all(delivererPromises);
+        // Find the current charter, and corresponding patentPolicy.
```
Comment (deniak) [inline lib/validator.js:541]: The comment seems inaccurate as we aren't dealing with any patent policy

**Review by deniak — CHANGES_REQUESTED** [review]

**Review by deniak — APPROVED** [review]

**Pull request merged**